소단락(새끼) 문제 목록 페이지 구현 완료

tails5555 opened the pull request:
- 상단부에 Breadthumb 와 문제 간략 정보를 조회할 수 있도록 구현함.
- 문제 목록에서 정보를 다르게 뒀음. 시험 이전이면 시험 시각, 이후면 제출 인원을 표기했음.
- 버튼 색상을 제출 기한에 따라 다르게 설정하였음.
- 데이터를 서버에서 받아옴을 가정하에 구현해서 서버 파트와 조율을 쉽게 할 수 있을 듯함.
- 미리보기는 카톡방을 참고하시길 바랍니다.

tails5555 committed:
#2 Init Main Title Component
메인 문제 목록 구현 중 상단부 구현 완료. 날짜 데이터는 localeString() 으로 가정하였음.

tails5555 committed:
#4 Update Main Title Component : Add Exam Type View
서버에서 받아온 시험 종류를 보여주는 것을 Vue 의 computed 를 사용하여 보여줌. 그리고 소단락(새끼) 문제들을 보여주는 UI 디자인 중.

tails5555 committed:
#4 Init TopBreadthumb Component
문제 목록 상단부에 Breadthumb 표기하는 기능을 구현함. Pivot 는 과목 이름과 대문제 제목으로 작성하겠음.

tails5555 committed:
#4 Init MainElement Component
소단락(새끼) 문제에 대하여 보여주는 기능 구현 중.
색상 조정 기능까지 구현하였음. 시간에 따른 기능과 데이터 Collapse 기능은 계속 구현하겠음.

tails5555 committed:
Merge branch 'master' of https://github.com/711skhu/front_core into feature-v1.0.2-main-problem-unit

tails5555 committed:
#4 Add Access Button
실습, 과제, 구현셤 별로 디테일한 접속 버튼 생성 완료.
시험 시간은 90분으로 잡아서 계산했음.

shouwn commented on src/components/main_problem/problem/MainElement.vue:
```diff
@@ -0,0 +1,90 @@
+            </sui-header-subheader>
+        </h2>
+        <h2 is="sui-header" floated="right" text-align="center">
+            <div v-if="new Date() >= problem.reserve_date">
```
위에도 ``` v-if="new Date() >= problem.reserve_date">```가 있는데 이걸 computed로 빼는 것도 방법일 것 같네요.

tails5555 replied:
computed 로 빼겠습니다. 좋은 방법이네요!

shouwn commented on src/components/main_problem/problem/MainElement.vue:
```diff
@@ -0,0 +1,90 @@
+            </sui-header-subheader>
+        </h2>
+    </sui-segment>
+</template>
```
로직을 보면 전체적으로  v-if="new Date() >= problem.reserve_date" 조건에 의해서 보여지는 내용이 달라지는데요. 

지금은 하나의 태그에 각 if 조건마다 어떻게 보여질지 다 적혀 있습니다.
이 부분을 2개의 자식 컴포넌트를 만들어서 v-if 에 따라 보여질 컴포넌트를 취사선택하고 전체 데이터를 넘겨주는 방식을 취하는 것도 좋을 것 같아요.

computed에도 이와 관련된 로직을 현재 이 컴포넌트 하나가 관리하고 있어서 무거운 느낌이 듭니다.

tails5555 replied:
Children API 를 사용하려 했지만, 2개의 컴포넌트를 따로 만들어 조건부 렌더링 처리를 최소화했습니다.

shouwn replied:
수정하신 detail-unit에서 checked 를 통해 detail-unit이 외부에 파라미터에 따라 자신이 보여줘야하는 것을 취사 선택하는 로직이 들어있습니다. 이는 detail-unit이 하는 일을 detail-unit을 사용하는 컴포넌트가 알아야하고, detail-unit은 외부에서 하는 일을 알아야 작동하는 구조로 두 컴포넌트의 관계가 매우 깊습니다.

그렇기 때문에 detail-unit은 props로 들어온 데이터를 어떻게 보여줄지에 대한 역할만 담당하고 무엇을 보여줄지를 MainElemnet에서 관리하는 것이 맞다고 생각합니다.

[조건부 렌더링](https://kr.vuejs.org/v2/guide/conditional.html#key%EB%A5%BC-%EC%9D%B4%EC%9A%A9%ED%95%9C-%EC%9E%AC%EC%82%AC%EC%9A%A9-%EA%B0%80%EB%8A%A5%ED%95%9C-%EC%97%98%EB%A6%AC%EB%A8%BC%ED%8A%B8-%EC%A0%9C%EC%96%B4)을 참고하면 좋을 것 같습니다.

tails5555 replied:
리뷰할 때 알 수 있는 단어로 작성 부탁 드립네다.

취사 선택은 뭐죠? 밥 짓는 건가요? ㅋㅋㅋ 국어사전 겨우 검색해서 이해했습니다 ㅋㅋㅋ

shouwn commented on src/components/main_problem/problem/MainTitle.vue:
```diff
@@ -0,0 +1,53 @@
+                    <sui-icon name="bell outline" />
+                    <sui-header-content>제출기한</sui-header-content>
+                </h1>
+                <p>{{ limit_date }}</p>
```
각 문제마다 제출 기한이 다른데 여기서 보여지는 limit date는 어떤 걸 나타내는 건가요?

tails5555 replied:
아 이를 특이사항 내용으로 바꾸겠습니다. 이는 오타입니다.

shouwn commented on src/components/main_problem/problem/MainTitle.vue:
```diff
@@ -0,0 +1,53 @@
+                    return '구현 실습';
+                default :
+                    return '기타 항목';
+            }
```
type에 따라 어떤 내용을 표시할지를 MainTitle 컴포넌트에서 정하는 것이 아니라 ExamType 클래스를 하나 만들어서 정의한 뒤에 거기에 title로 프로퍼티를 만드는 방법도 좋을 것 같습니다.

tails5555 replied:
알겠습니다. 이는 객체 안에서 기재하겠습니다.

tails5555 committed:
#4 Update About Review Constraints
코드 리뷰에서 남긴 내용에 따라 4개 중 3개 수정함. 나머지 하나는 더 알아보고 수정하겠음.

tails5555 committed:
#4 Separate Score, Detail Unit
코드 리뷰에서 지적한 사항 중에 조건 렌더링에 대한 부분 수정 및 점수 시각화의 미흡 사항을 보안하였음.

jihyuno301 commented on src/components/main_problem/problem/MainTitle.vue:
```diff
@@ -6,17 +6,16 @@
             <sui-grid-column width="eight">
                 <h1 is="sui-header">
                     <sui-icon name="bell outline" />
-                    <sui-header-content>제출기한</sui-header-content>
```
각 세그먼트마다 조건부로 제출기한 보여줘서 삭제한건가요?

tails5555 replied:
예 그렇습니다. 제출기한은 문제 별로 다르게 책정되어 있습니다.
그래서 Main Title 컴포넌트에 제출기한을 보여주지 않고 특이사항으로 대체했습니다.

shouwn commented on src/components/main_problem/problem/MainTitle.vue:
```diff
@@ -6,17 +6,16 @@
+                    <sui-header-content>특이사항</sui-header-content>
                 </h1>
-                <p>{{ limit_date }}</p>
+                <p>{{ context }}</p>
```
특이사항의 내용 props가 context인 것은 이 컴포넌트를 사용하는 외부 사용자가 무엇을 나타내는지 헷갈릴 수 있을 것 같습니다. 이름을 바꾸는 게 좋을 것 같네요.

tails5555 replied:
그러면 이름을 `description` 로 바꿀게요.

shouwn commented:
현재 props로 들어온 모든 변수들에 대한 상세한 명시가 없습니다.
이 부분은 Vue.js 컨벤션에도 적혀 있는 내용이기 때문에 상세히 적어주셨으면 합니다.

[props 컨벤션](https://kr.vuejs.org/v2/style-guide/index.html#Props-%EC%A0%95%EC%9D%98-%ED%95%84%EC%88%98)

shouwn reviewed (changes requested):
현재 props로 들어온 모든 변수들에 대한 상세한 명시가 없습니다.
이 부분은 Vue.js 컨벤션에도 적혀 있는 내용이기 때문에 상세히 적어주셨으면 합니다.

[props 컨벤션](https://kr.vuejs.org/v2/style-guide/index.html#Props-%EC%A0%95%EC%9D%98-%ED%95%84%EC%88%98)

tails5555 committed:
#4 Update About Props Convention
Props 속성 변수 타입과 validate 속성 추가. 그리고 애매한 변수 이름 수정과 보기 안 좋은 로직을 다듬었음. 객체 타입에서 Object 는 서버에서 보내지는 데이터 타입이 달라질 수 있어 이로 지정함.

tails5555 commented:
실수로 closed 버튼을 클릭했군요. 양해 부탁 드립니다.

그리고 요구했던 props 컨벤션 정리를 했습니다.

tails5555 committed:
#4 Update About Component Props Rendering
Props 를 전달하는 것만 수정하였음. 자식 컴포넌트 안에선 조건부 렌더링을 진행하지 않아도 됨.
